In [178]:
from google.colab import drive # Import the drive module from google.colab to interact with Google Drive
import os # Import the os module for interacting with the operating system, like changing directories
drive.mount('/content/drive') # Mount Google Drive to the Colab environment at '/content/drive'
os.chdir('/content/drive/MyDrive/Academics/Visiting Lectures/2026-H1/202601-SDP-AU/Session-08-Data-Management-in-Pandas') # Change the current working directory to the specified path in Google Drive

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [179]:
import pandas as pd # Import the pandas library, commonly used for data manipulation and analysis
import os # Import the os module for interacting with the operating system, e.g., path operations

folder_name = 'Data' # Define the name of the folder where the data file is located
excel_file_path = os.path.join(folder_name, 'bank_marketing_data.xlsx') # Construct the full path to the Excel file using os.path.join for cross-platform compatibility

# Check if the 'Data' folder exists (this comment was already present, keeping it for context)
bank_data = pd.read_excel(excel_file_path) # Read the Excel file into a pandas DataFrame named 'bank_data'

## **Data Frame Inspection: `Head()`**

In [180]:
bank_data.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

The output above shows the first 5 rows of the `bank_data` DataFrame, providing a glimpse into its structure and content. Each row represents a client, and the columns provide various attributes related to them and the marketing campaign:

-   **age**: Age of the client.
-   **job**: Type of job of the client.
-   **marital**: Marital status of the client.
-   **education**: Educational background of the client.
-   **default**: Has credit in default? ('yes', 'no', 'unknown').
-   **housing**: Has housing loan? ('yes', 'no', 'unknown').
-   **loan**: Has personal loan? ('yes', 'no', 'unknown').
-   **contact**: Contact communication type ('cellular', 'telephone').
-   **month**: Last contact month of the year.
-   **day_of_week**: Last contact day of the week.
-   **campaign**: Number of contacts performed during this campaign for this client.
-   **pdays**: Number of days that passed by after the client was last contacted from a previous campaign (999 means client was not previously contacted).
-   **previous**: Number of contacts performed before this campaign and for this client.
-   **poutcome**: Outcome of the previous marketing campaign ('failure', 'nonexistent', 'success').
-   **emp.var.rate**: Employment variation rate (quarterly indicator).
-   **cons.price.idx**: Consumer price index (monthly indicator).
-   **cons.conf.idx**: Consumer confidence index (monthly indicator).
-   **euribor3m**: Euribor 3 month rate (daily indicator).
-   **nr.employed**: Number of employees (quarterly indicator).
-   **y**: Has the client subscribed to a term deposit? ('yes', 'no').

This initial inspection helps to understand the types of data (numerical, categorical) and potential issues like 'unknown' values, which might require further cleaning or imputation.

## **Data Frame Inspection: `Info()`**



In [181]:
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

## **Data Frame Inspection: `Shape`**



In [182]:
display(Markdown(f"### The bank_data DataFrame has {bank_data.shape[0]} rows and {bank_data.shape[1]} columns."))

### The bank_data DataFrame has 41188 rows and 21 columns.

## **Data Frame Inspection: `Describe()`**


In [183]:
bank_data.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

## **Data Frame Inspection: `Value_Counts()` (Categorical)**



In [184]:
bank_data['job'].value_counts()

job
admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: count, dtype: int64

## **Missing Values: Check Explicit NaNs**
Calculate and display the sum of explicit missing values (NaNs) for each column in the `bank_data` DataFrame using `isnull().sum()`.


In [185]:
bank_data.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

## **Missing Values: Identify 'unknown' in Categorical Columns**
Identify and quantify implicit missing values represented as 'unknown' strings within categorical columns. This step will list columns containing 'unknown' and their respective counts and percentages.


In [186]:
from IPython.display import display, Markdown # Import display and Markdown

unknown_data = [] # Initialize an empty list to store information about 'unknown' values
categorical_cols_with_unknown = ['job', 'marital', 'education', 'default', 'housing', 'loan'] # Define the categorical columns to check for 'unknown' values

# Iterate through each specified categorical column
for col in categorical_cols_with_unknown:
    unknown_count = bank_data[col].astype(str).str.lower().eq('unknown').sum() # Count occurrences of 'unknown' (case-insensitive)
    total_rows = len(bank_data) # Get the total number of rows in the DataFrame

    # Check if 'unknown' values exist in the current column
    if unknown_count > 0:
        percentage = (unknown_count / total_rows) * 100 # Calculate the percentage of 'unknown' values
        unknown_data.append({
            'Column': col,
            'Unknown Count': unknown_count,
            'Percentage': f'{percentage:.2f}%' # Format percentage to two decimal places
        }) # Append the findings to the unknown_data list

# Display the aggregated information about 'unknown' values
display(Markdown("### Summary of 'unknown' values in categorical columns:"))
for item in unknown_data:
    display(f"Column: {item['Column']}, Unknown Count: {item['Unknown Count']}, Percentage: {item['Percentage']}")

### Summary of 'unknown' values in categorical columns:

'Column: job, Unknown Count: 330, Percentage: 0.80%'

'Column: marital, Unknown Count: 80, Percentage: 0.19%'

'Column: education, Unknown Count: 1731, Percentage: 4.20%'

'Column: default, Unknown Count: 8597, Percentage: 20.87%'

'Column: housing, Unknown Count: 990, Percentage: 2.40%'

'Column: loan, Unknown Count: 990, Percentage: 2.40%'

## **Missing Values: Impute 'unknown' in 'default'**
Replace 'unknown' values in the 'default' column with 'no', as per business understanding that 'unknown' often implies no default. This improves data quality for this specific column.


In [187]:
bank_data['default'] = bank_data['default'].replace('unknown', 'no')
display(Markdown("### Values in 'default' column after imputation:"))
display(bank_data['default'].value_counts())

### Values in 'default' column after imputation:

default
no     41185
yes        3
Name: count, dtype: int64

## **Missing Values: Impute 'unknown' with Mode**
For other categorical columns containing 'unknown' values (excluding 'default'), replace 'unknown' with the mode (most frequent value) of each respective column. This provides a data-driven imputation strategy.


In [188]:
columns_to_impute_with_mode = ['job', 'marital', 'education', 'housing', 'loan'] # Define the list of columns for mode imputation

display(Markdown("### Imputing 'unknown' values with mode for specified columns:"))
for col in columns_to_impute_with_mode:
    mode_value = bank_data[col].mode()[0] # Calculate the mode (most frequent value) of the column
    bank_data[col] = bank_data[col].replace('unknown', mode_value) # Replace 'unknown' with the calculated mode
    display(Markdown(f"#### Column '{col}' after imputation:"))
    display(bank_data[col].value_counts()) # Display value counts to verify the imputation

### Imputing 'unknown' values with mode for specified columns:

#### Column 'job' after imputation:

job
admin.           10752
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
Name: count, dtype: int64

#### Column 'marital' after imputation:

marital
married     25008
single      11568
divorced     4612
Name: count, dtype: int64

#### Column 'education' after imputation:

education
university.degree      13899
high.school             9515
basic.9y                6045
professional.course     5243
basic.4y                4176
basic.6y                2292
illiterate                18
Name: count, dtype: int64

#### Column 'housing' after imputation:

housing
yes    22566
no     18622
Name: count, dtype: int64

#### Column 'loan' after imputation:

loan
no     34940
yes     6248
Name: count, dtype: int64

In [189]:
verification_results = [] # Initialize an empty list to store verification results

# Iterate through the columns that were supposed to be imputed
for col in columns_to_impute_with_mode:
    remaining_unknown_count = bank_data[col].astype(str).str.lower().eq('unknown').sum() # Count any remaining 'unknown' (case-insensitive)
    verification_results.append({
        'Column': col,
        'Unknown Count Remaining': remaining_unknown_count
    }) # Append the result for the current column

display(Markdown("### Verification of 'unknown' values after imputation:"))
for item in verification_results:
    display(f"Column: {item['Column']}, Unknown Count Remaining: {item['Unknown Count Remaining']}")

### Verification of 'unknown' values after imputation:

'Column: job, Unknown Count Remaining: 0'

'Column: marital, Unknown Count Remaining: 0'

'Column: education, Unknown Count Remaining: 0'

'Column: housing, Unknown Count Remaining: 0'

'Column: loan, Unknown Count Remaining: 0'

## **Selection: Select a Single Column**
Demonstrate how to select a single column from the `bank_data` DataFrame using bracket notation and display its first few entries. This is a fundamental selection method.


In [190]:
bank_data['age'].head()

0    56
1    57
2    37
3    40
4    56
Name: age, dtype: int64

## **Selection: Select Multiple Columns**
Demonstrate how to select multiple specific columns from the `bank_data` DataFrame using a list of column names and display their first few entries.


In [191]:
selected_columns = ['age', 'job', 'marital'] # Define a list of columns to select
bank_data[selected_columns].head() # Select the columns and display the first 5 rows

age        job  marital
0   56  housemaid  married
1   57   services  married
2   37   services  married
3   40     admin.  married
4   56   services  married

## **Selection: Select Rows by Label (.loc)**
Illustrate selecting rows by label using the `.loc` accessor, for example, selecting the first 5 rows. This highlights label-based indexing capabilities.


In [192]:
bank_data.loc[0:4]

age        job  marital    education default housing loan    contact month  \
0   56  housemaid  married     basic.4y      no      no   no  telephone   may   
1   57   services  married  high.school      no      no   no  telephone   may   
2   37   services  married  high.school      no     yes   no  telephone   may   
3   40     admin.  married     basic.6y      no      no   no  telephone   may   
4   56   services  married  high.school      no      no  yes  telephone   may   

  day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0         mon  ...         1    999         0  nonexistent          1.1   
1         mon  ...         1    999         0  nonexistent          1.1   
2         mon  ...         1    999         0  nonexistent          1.1   
3         mon  ...         1    999         0  nonexistent          1.1   
4         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

## **Selection: Select Rows by Integer Position (.iloc)**
Show how to select rows by integer position using the `.iloc` accessor, for example, selecting rows from index 10 to 14. This demonstrates positional indexing.


In [193]:
bank_data.iloc[10:15]

age          job   marital          education default housing loan  \
10   41  blue-collar   married  university.degree      no      no   no   
11   25     services    single        high.school      no     yes   no   
12   29  blue-collar    single        high.school      no      no  yes   
13   57    housemaid  divorced           basic.4y      no     yes   no   
14   35  blue-collar   married           basic.6y      no     yes   no   

      contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
10  telephone   may         mon  ...         1    999         0  nonexistent   
11  telephone   may         mon  ...         1    999         0  nonexistent   
12  telephone   may         mon  ...         1    999         0  nonexistent   
13  telephone   may         mon  ...         1    999         0  nonexistent   
14  telephone   may         mon  ...         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
10          1.1          93.994          -36.4      4.857       5191.0  no  
11          1.1          93.994          -36.4      4.857       5191.0  no  
12          1.1          93.994          -36.4      4.857       5191.0  no  
13          1.1          93.994          -36.4      4.857       5191.0  no  
14          1.1          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

## **Selection: Filter Rows by Single Condition**
Filter the `bank_data` DataFrame to select rows where a specific condition is met, such as 'age' greater than 60, and display the head of the filtered DataFrame.


In [194]:
filtered_data_age_60_plus = bank_data[bank_data['age'] > 60]
display(Markdown("### First 5 rows of bank_data where age > 60:"))
display(filtered_data_age_60_plus.head())

### First 5 rows of bank_data where age > 60:

age         job   marital            education default housing loan  \
8645    61     retired   married          high.school      no     yes   no   
27713   70     retired  divorced             basic.4y      no     yes   no   
27715   61      admin.   married    university.degree      no     yes   no   
27723   66  technician   married  professional.course      no     yes   no   
27757   76     retired   married    university.degree      no      no  yes   

         contact month day_of_week  ...  campaign  pdays  previous  \
8645   telephone   jun         wed  ...         4    999         0   
27713   cellular   mar         mon  ...         3    999         0   
27715   cellular   mar         mon  ...         6    999         0   
27723   cellular   mar         tue  ...         1    999         0   
27757   cellular   mar         thu  ...         9    999         0   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
8645   nonexistent          1.4          94.465          -41.8      4.864   
27713  nonexistent         -1.8          92.843          -50.0      1.811   
27715  nonexistent         -1.8          92.843          -50.0      1.811   
27723  nonexistent         -1.8          92.843          -50.0      1.799   
27757  nonexistent         -1.8          92.843          -50.0      1.757   

       nr.employed    y  
8645        5228.1   no  
27713       5099.1  yes  
27715       5099.1   no  
27723       5099.1  yes  
27757       5099.1   no  

[5 rows x 21 columns]

## **Selection: Filter Rows by Multiple Conditions**
Filter the `bank_data` DataFrame based on multiple combined conditions (e.g., 'marital' is 'married' AND 'education' is 'university.degree' AND 'age' between 30 and 50). This demonstrates complex row selection for targeted segments.


In [195]:
condition_marital = bank_data['marital'] == 'married'
condition_education = bank_data['education'] == 'university.degree'
condition_age = (bank_data['age'] >= 30) & (bank_data['age'] <= 50)

# Combine all conditions using logical AND
combined_condition = condition_marital & condition_education & condition_age

# Apply the combined condition to filter the DataFrame
targeted_segment = bank_data[combined_condition]

display(Markdown("### First 5 rows of the targeted_segment DataFrame (married, university.degree, age 30-50):"))
display(targeted_segment.head())

### First 5 rows of the targeted_segment DataFrame (married, university.degree, age 30-50):

age          job  marital          education default housing loan  \
7    41  blue-collar  married  university.degree      no      no   no   
10   41  blue-collar  married  university.degree      no      no   no   
25   35   technician  married  university.degree      no      no  yes   
30   46       admin.  married  university.degree      no      no   no   
32   49  blue-collar  married  university.degree      no      no   no   

      contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
7   telephone   may         mon  ...         1    999         0  nonexistent   
10  telephone   may         mon  ...         1    999         0  nonexistent   
25  telephone   may         mon  ...         1    999         0  nonexistent   
30  telephone   may         mon  ...         1    999         0  nonexistent   
32  telephone   may         mon  ...         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
7           1.1          93.994          -36.4      4.857       5191.0  no  
10          1.1          93.994          -36.4      4.857       5191.0  no  
25          1.1          93.994          -36.4      4.857       5191.0  no  
30          1.1          93.994          -36.4      4.857       5191.0  no  
32          1.1          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

## **Selection: Select Specific Rows and Columns**
Combine row and column selection using `.loc` to select specific columns for a filtered set of rows (e.g., 'age' and 'job' for married university graduates between 30 and 50).


In [196]:
selected_rows_and_cols = targeted_segment.loc[:, ['age', 'job']]
display(Markdown("### First 5 rows of 'age' and 'job' columns for married university graduates between 30 and 50:"))
display(selected_rows_and_cols.head())

### First 5 rows of 'age' and 'job' columns for married university graduates between 30 and 50:

age          job
7    41  blue-collar
10   41  blue-collar
25   35   technician
30   46       admin.
32   49  blue-collar

## **Variable Creation: Create 'age_group'**
Create a new categorical column named 'age_group' in `bank_data` by categorizing the 'age' column into 'young', 'middle-aged', and 'senior' groups using `pd.cut()`. This adds a new feature for demographic analysis.


In [197]:
age_bins = [0, 30, 50, 100] # Define the boundaries for age groups
age_labels = ['young', 'middle-aged', 'senior'] # Define the labels for the age groups

bank_data['age_group'] = pd.cut(bank_data['age'], bins=age_bins, labels=age_labels, right=False) # Create the 'age_group' column using pd.cut()

display(Markdown("### First 5 rows of bank_data with the new 'age_group' column:"))
display(bank_data.head())

### First 5 rows of bank_data with the new 'age_group' column:

age        job  marital    education default housing loan    contact month  \
0   56  housemaid  married     basic.4y      no      no   no  telephone   may   
1   57   services  married  high.school      no      no   no  telephone   may   
2   37   services  married  high.school      no     yes   no  telephone   may   
3   40     admin.  married     basic.6y      no      no   no  telephone   may   
4   56   services  married  high.school      no      no  yes  telephone   may   

  day_of_week  ...  pdays  previous     poutcome  emp.var.rate cons.price.idx  \
0         mon  ...    999         0  nonexistent           1.1         93.994   
1         mon  ...    999         0  nonexistent           1.1         93.994   
2         mon  ...    999         0  nonexistent           1.1         93.994   
3         mon  ...    999         0  nonexistent           1.1         93.994   
4         mon  ...    999         0  nonexistent           1.1         93.994   

   cons.conf.idx  euribor3m  nr.employed   y    age_group  
0          -36.4      4.857       5191.0  no       senior  
1          -36.4      4.857       5191.0  no       senior  
2          -36.4      4.857       5191.0  no  middle-aged  
3          -36.4      4.857       5191.0  no  middle-aged  
4          -36.4      4.857       5191.0  no       senior  

[5 rows x 22 columns]

## **Variable Creation: Verify 'age_group'**
Display the value counts of the newly created 'age_group' column to verify the distribution across different age categories.


In [198]:
display(Markdown("### Value counts for 'age_group' column:"))
display(bank_data['age_group'].value_counts())

### Value counts for 'age_group' column:

age_group
middle-aged    27464
senior          8055
young           5669
Name: count, dtype: int64

## **Variable Creation: Convert Target Variable 'y' to Numerical**
Convert the target variable 'y' from categorical ('yes', 'no') to numerical (1, 0) for machine learning readiness. Display the value counts of the converted column.


In [199]:
bank_data['y'] = bank_data['y'].replace({'yes': 1, 'no': 0})
print("Value counts for the converted 'y' column:")
bank_data['y'].value_counts()

Value counts for the converted 'y' column:


/tmp/ipython-input-56987362.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  bank_data['y'] = bank_data['y'].replace({'yes': 1, 'no': 0})


y
0    36548
1     4640
Name: count, dtype: int64

In [200]:
bank_data['y'] = bank_data['y'].replace({'yes': 1, 'no': 0}).astype(int)
display(Markdown("### Value counts for the converted 'y' column:"))
display(bank_data['y'].value_counts())

### Value counts for the converted 'y' column:

y
0    36548
1     4640
Name: count, dtype: int64

## **Variable Creation: Handle 'pdays' Column**
Replace the special value '999' in the 'pdays' column with 'never_contacted' to make its meaning explicit, and then convert the column to a categorical data type. This improves interpretability and data consistency.


In [201]:
bank_data['pdays'] = bank_data['pdays'].replace(999, 'never_contacted') # Replace 999 with 'never_contacted'
bank_data['pdays'] = bank_data['pdays'].astype('category') # Convert to categorical type

display(Markdown("### Value counts for 'pdays' column after handling '999' and converting to categorical:"))
display(bank_data['pdays'].value_counts())

### Value counts for 'pdays' column after handling '999' and converting to categorical:

pdays
never_contacted    39673
3                    439
6                    412
4                    118
9                     64
2                     61
7                     60
12                    58
10                    52
5                     46
13                    36
11                    28
1                     26
15                    24
14                    20
8                     18
0                     15
16                    11
17                     8
18                     7
19                     3
22                     3
21                     2
20                     1
25                     1
26                     1
27                     1
Name: count, dtype: int64

## **Variable Creation: Verify Data Types**
Display the data types of the 'y' and 'pdays' columns after conversion to ensure they are consistent with the desired numerical and categorical types.


In [202]:
display(Markdown("### Data types of 'y' and 'pdays' columns:"))
display(bank_data[['y', 'pdays']].dtypes)

### Data types of 'y' and 'pdays' columns:

y           int64
pdays    category
dtype: object

## **Data Aggregation: Job Performance**
Aggregate the `bank_data` by 'job' to calculate metrics like mean contact duration, total campaign contacts, successful subscriptions, and a derived subscription rate. This helps in analyzing performance across different job sectors.


In [203]:
job_performance = bank_data.groupby('job').agg(
    mean_duration=('duration', 'mean'),
    total_campaign_contacts=('campaign', 'sum'),
    successful_subscriptions=('y', 'sum'),
    total_contacts=('y', 'count')
).reset_index()

job_performance['subscription_rate'] = (job_performance['successful_subscriptions'] / job_performance['total_contacts']) * 100

display(Markdown("### First 5 rows of the job_performance DataFrame:"))
display(job_performance.head())

### First 5 rows of the job_performance DataFrame:

job  mean_duration  total_campaign_contacts  \
0        admin.     253.862909                    28216   
1   blue-collar     264.542360                    23676   
2  entrepreneur     263.267857                     3692   
3     housemaid     250.454717                     2798   
4    management     257.058140                     7240   

   successful_subscriptions  total_contacts  subscription_rate  
0                      1389           10752          12.918527  
1                       638            9254           6.894316  
2                       124            1456           8.516484  
3                       106            1060          10.000000  
4                       328            2924          11.217510

## **Data Aggregation: Display Job Performance**
Display the aggregated `job_performance` DataFrame, sorted by subscription rate, to highlight which job categories are most responsive to marketing campaigns.


In [204]:
job_performance_sorted = job_performance.sort_values(by='subscription_rate', ascending=False)
display(Markdown("### Job performance sorted by subscription rate (highest first):"))
display(job_performance_sorted)

### Job performance sorted by subscription rate (highest first):

job  mean_duration  total_campaign_contacts  \
8         student     283.683429                     1841   
5         retired     273.712209                     4260   
10     unemployed     249.451677                     2600   
0          admin.     253.862909                    28216   
4      management     257.058140                     7240   
9      technician     250.232241                    17379   
6   self-employed     264.142153                     3781   
3       housemaid     250.454717                     2798   
2    entrepreneur     263.267857                     3692   
7        services     258.398085                    10271   
1     blue-collar     264.542360                    23676   

    successful_subscriptions  total_contacts  subscription_rate  
8                        275             875          31.428571  
5                        434            1720          25.232558  
10                       144            1014          14.201183  
0                       1389           10752          12.918527  
4                        328            2924          11.217510  
9                        730            6743          10.826042  
6                        149            1421          10.485574  
3                        106            1060          10.000000  
2                        124            1456           8.516484  
7                        323            3969           8.138070  
1                        638            9254           6.894316

## **Data Aggregation: Education Performance**
Aggregate the `bank_data` by 'education' to calculate mean contact duration, total campaign contacts, successful subscriptions, and a derived subscription rate. This provides insights into campaign effectiveness based on education levels.


In [205]:
education_performance = bank_data.groupby('education').agg(
    mean_duration=('duration', 'mean'),
    total_campaign_contacts=('campaign', 'sum'),
    successful_subscriptions=('y', 'sum'),
    total_contacts=('y', 'count')
).reset_index()

education_performance['subscription_rate'] = (education_performance['successful_subscriptions'] / education_performance['total_contacts']) * 100

display(Markdown("### First 5 rows of the education_performance DataFrame:"))
display(education_performance.head())

### First 5 rows of the education_performance DataFrame:

education  mean_duration  total_campaign_contacts  \
0     basic.4y     264.827826                    10860   
1     basic.6y     264.414485                     5859   
2     basic.9y     261.291811                    15308   
3  high.school     260.886810                    24440   
4   illiterate     276.777778                       41   

   successful_subscriptions  total_contacts  subscription_rate  
0                       428            4176          10.249042  
1                       188            2292           8.202443  
2                       473            6045           7.824648  
3                      1031            9515          10.835523  
4                         4              18          22.222222

In [206]:
education_performance_sorted = education_performance.sort_values(by='subscription_rate', ascending=False)
display(Markdown("### Education performance sorted by subscription rate (highest first):"))
display(education_performance_sorted)

### Education performance sorted by subscription rate (highest first):

education  mean_duration  total_campaign_contacts  \
4           illiterate     276.777778                       41   
6    university.degree     254.365062                    35687   
5  professional.course     252.533855                    13559   
3          high.school     260.886810                    24440   
0             basic.4y     264.827826                    10860   
1             basic.6y     264.414485                     5859   
2             basic.9y     261.291811                    15308   

   successful_subscriptions  total_contacts  subscription_rate  
4                         4              18          22.222222  
6                      1921           13899          13.821138  
5                       595            5243          11.348465  
3                      1031            9515          10.835523  
0                       428            4176          10.249042  
1                       188            2292           8.202443  
2                       473            6045           7.824648

## **Summary:**

The data management activities performed in this tutorial primarily involved data inspection, handling of implicit missing values, feature engineering (variable creation), and data aggregation. Their importance in preparing data for analysis and driving business insights can be explained as follows:

1.  **Data Inspection (Head, Info, Shape, Describe, Value Counts):** This initial phase provides a foundational understanding of the dataset's structure, content, data types, dimensions (41,188 rows, 21 columns), and statistical properties. It helps identify potential issues early, such as non-standard missing value representations (e.g., 'unknown', 999) and data distributions (e.g., the right-skewed `duration` column with a mean of 258.29 seconds but a median of 180 seconds). This step is crucial for comprehending the raw data's characteristics and informing subsequent cleaning strategies.

2.  **Missing Value Handling:**
    *   **Identifying Explicit NaNs:** Confirming the absence of explicit `NaN` values ($0$ for all columns) simplifies initial data cleaning but highlights the need to look for other missing value representations.
    *   **Identifying and Imputing 'unknown' values:** The presence of 'unknown' in several categorical columns (e.g., $20.87\%$ in 'default', $4.20\%$ in 'education') indicates implicit missing data. Replacing 'unknown' with 'no' in 'default' (based on business logic) and with the mode in other columns (like 'job', 'marital', 'education', 'housing', 'loan') ensures that these entries are not treated as valid categories but rather as known or imputed values. This step cleans the data, preventing skewed analysis results and improving model performance.

3.  **Feature Engineering (Variable Creation):**
    *   **Creating 'age\_group':** Categorizing 'age' into 'young', 'middle-aged', and 'senior' creates a new, more interpretable feature for demographic analysis, allowing for high-level segmentation.
    *   **Converting 'y' to Numerical:** Transforming the target variable 'y' from 'yes'/'no' to $1$/$0$ makes it suitable for machine learning algorithms, which require numerical inputs.
    *   **Handling 'pdays':** Replacing '999' with 'never\_contacted' and converting it to a categorical type makes the column's meaning explicit and improves interpretability, as '999' is a special indicator rather than a numerical value.
    These steps enhance the dataset's analytical potential by creating new features or transforming existing ones into more suitable formats for specific analysis tasks or modeling.

4.  **Data Aggregation:**
    *   **Job Performance and Education Performance:** Aggregating data by 'job' and 'education' to calculate metrics like mean contact duration, total campaign contacts, successful subscriptions, and subscription rates helps identify key segments. For instance, students showed the highest subscription rate (approx. $31.43\%$), while 'illiterate' individuals had a high subscription rate ($22.22\%$) among education levels. This provides actionable business insights, enabling targeted marketing strategies and resource allocation.

In summary, these data management activities are critical because they transform raw, potentially messy data into a clean, well-structured, and enriched format. This preparation ensures data quality, facilitates meaningful analysis, uncovers actionable insights into customer behavior and campaign effectiveness, and ultimately drives informed business decisions.

### Data Analysis Key Findings
*   The initial dataset contains 41,188 entries and 21 columns, with no explicit missing values (NaNs).
*   Implicit missing values were identified as 'unknown' in several categorical columns, most notably in 'default' ($20.87\%$ or 8597 entries) and 'education' ($4.20\%$ or 1731 entries).
*   The 'unknown' values in the 'default' column were imputed with 'no', and in 'job', 'marital', 'education', 'housing', and 'loan' columns with their respective modes, ensuring cleaner and more consistent categorical data.
*   A new 'age\_group' column was created, categorizing ages into 'young' (5669 entries), 'middle-aged' (27464 entries), and 'senior' (8055 entries).
*   The target variable 'y' was converted to a numerical format ($1$ for 'yes', $0$ for 'no'), with 4640 successful subscriptions ($1$) and 36548 non-subscriptions ($0$).
*   The special value $999$ in 'pdays' (representing no prior contact) was replaced with 'never\_contacted' (39673 occurrences) and converted to a categorical type for better interpretability.
*   Analysis of job performance revealed that **students** had the highest subscription rate (approx. $31.43\%$), followed by **retired** individuals (approx. $25.23\%$), while the **blue-collar** job category had the lowest (approx. $6.89\%$).
*   Analysis of education performance showed that 'illiterate' individuals had the highest subscription rate ($22.22\%$), followed by 'university.degree' ($13.82\%$).

### Insights or Next Steps
*   **Targeted Campaign Strategy:** Focus marketing efforts on demographic segments with historically higher subscription rates, such as students and retired individuals, and tailor campaigns for those with 'illiterate' or 'university.degree' education backgrounds, as they show higher responsiveness.
*   **Further Explore 'unknown' Values:** While imputed, the high percentage of 'unknown' values in columns like 'default' ($20.87\%$) and 'education' ($4.20\%$) suggests potential data collection issues. Investigate the source of these 'unknown' entries to understand if there's a systemic problem or if a more sophisticated imputation method is warranted (e.g., predictive imputation based on other features).
